### PR TITLE
Feat/bivariate synth

### PR DESCRIPTION
**Pull request by SybelBlue**
tl;dr: My attempt at integrating the bivariate API functions `sjoin` and `merge` into the program scope.

Both functions range over all possible candidates:
- `merge` looks for column indices from separate frames that are `dtype`-equal 
- `sjoin` ranges over all pairs of GeoDataFrames
- both yield a program and its 'dual'
  - Because geopandas bifunctions have `(left, right, how)`, one would *assume* that duals are equivalent, ie `(l, r, h) == (r, l, invert(h))`
  - This is an *erroneous assumption*. Simply switching the ordering of bindings in the input dict results in failure. 
  - Therefore each synthesizer generates programs and their duals (until we can figure out something better)

Other big changes:
- All non-lazy code was removed following our previous chats
- Extra grammar terminals were added, and `GrammarRule` was renamed to `Candidate`
- Inputs to bivariate synthesizers must be `GdfBindings` which extend `dict[str, GeoDataFrame]` but add some useful memoization that should greatly streamline computation over `len(bindings) > 2`

### CODE REVIEW

File: src/grammar.py [hunk 1]
```diff
@@ -40,25 +42,50 @@ def interpret(self, gdfs):
...
 
 
 @dataclass(frozen=True, repr=False)
-class SJoin(GrammarRule):
+class SJoin(Candidate):
     left: str
     right: str
+    _: KW_ONLY
+    how: str
+    predicate: str
```
Comment (parkerziegler) [inline src/grammar.py]: To help us remember the literal values we support for parameters in our synthesizer, could we specify these types explicitly as their string literal unions, e.g.:

```suggestion
    how: Literal["left", "right", "inner"]
    predicate: Literal["intersects", "within", "contains", "overlaps", "crosses", "touches"]
```

File: src/synthesize.py [hunk 2]
```diff
@@ -1,114 +1,136 @@
...
+    for p in synth_matching(program, gdfs, target):
+        print('\t', p)
+        out.append(p)
+    print('Done.')
+    return out
+
+
+def binding_pairs(gdfs: GdfBindings):
+    'Returns a generator over all unique (unordered) pairs of DataFrames in gdfs'
+    return combinations(gdfs.items(), 2)
```
Comment (parkerziegler) [inline src/synthesize.py:100]: I may be misunderstanding here, but do we want `combinations` or `permutations` here? IIRC, `combinations` is lexicographically ordered based on the input ordering of the iterable, whereas `permutations` is unordered. I think this would only return a single ordering for a given dictionary of `gdfs` of length 2, rather than returning both.

File: src/grammar.py (hunk 1 above)
Comment (SybelBlue) [reply src/grammar.py]: Great idea!

File: src/synthesize.py (hunk 2 above)
Comment (SybelBlue) [reply src/synthesize.py:100]: You are correct! Instead of using two `yield` statements in `sjoin` and `merge` to get both orderings of each combination, we could use `permutations` instead of `combinations`. 

I wrote it this way because I thought it was conceptually easier for `binding_pairs` to produce each pair of things exactly one time, and let callers handle the ordering problem. 

At the time, I also wasn't sure if `merge` was order-sensitive but I'm pretty sure it is, may be worth following up with @ethanlee16 on that.

File: src/example.py [hunk 3]
```diff
@@ -21,36 +26,42 @@
...
-#   with    dataclass and no print: ~8.8 secs
-# benchmark('synthesize(input_gdfs, target)')
-
-# test code:
-#   lazy_synthesize(input_gdfs, target)
-# completed in:
-#   without dataclass optimization: ~0.36 secs
-#   with    dataclass optimization: ~0.35 secs
-benchmark('lazy_synthesize(input_gdfs, target)')
+    print('\n'.join(f.getvalue().splitlines()[:8]))
```
Comment (parkerziegler) [inline src/example.py:45]: Nit — let's update the comment above since we truncate to 8 lines here!

File: src/synthesize.py [hunk 4]
```diff
@@ -1,114 +1,136 @@
...
-        for col in gdf.columns:
-            if col != 'geometry':
-                yield Dissolve(GDF(gdf_name), col)
 
-
-def make_candidate_filter(gdfs: dict[str, GeoDataFrame], target: GeoDataFrame):
+def make_candidate_filter(gdfs: GdfBindings, target: DataFrame) -> Callable[[Candidate], bool]:
     'Returns a predicate that checks if a `program` over `gdfs` matches `target`'
-    return lambda program: gdfs_equal(evaluate_program(program, gdfs), target)
+    return lambda program: GeoDataFrame.equals(program.interpret(gdfs), target)
```
Comment (parkerziegler) [inline src/synthesize.py]: I think we'll still want to use [`assert_geodataframe_equal`](https://geopandas.org/en/stable/docs/reference/api/geopandas.testing.assert_geodataframe_equal.html#geopandas.testing.assert_geodataframe_equal) from `geopandas` for dataframe equivalence checking. A big reason for this is the `check_crs` parameter, which will verify that the `GeoDataFrame` produced by our program is in the same coordinate reference system as the output. This will be important when we add synthesis support for reprojection, which is an incredibly common spatial analysis task but one participants forget to do often!

File: src/synthesize.py (hunk 2 above)
Comment (parkerziegler) [reply src/synthesize.py:100]: Ah gotcha, sounds good!

**Review by parkerziegler — APPROVED** [review]
Looks great @SybelBlue! Merge when you like, but please remember to squash the commits 😄

File: src/synthesize.py (hunk 4 above)
Comment (SybelBlue) [reply src/synthesize.py]: Gotcha. I wonder if there's a way around using an assert because the creation of try/except blocks is big performance hit (about +25%). Reading the geopandas source [here](https://github.com/geopandas/geopandas/blob/be647ab7e54880278cba374cc77075afd33b25a5/geopandas/testing.py#L239), it looks like we'd need to write our own version of it. What do you think?

Comment (parkerziegler) [reply src/synthesize.py]: Ah gotcha! In that case, I would just steal and adapt [these lines](https://github.com/geopandas/geopandas/blob/be647ab7e54880278cba374cc77075afd33b25a5/geopandas/testing.py#L292-L295) checking the CRS equivalence. In essence, we _never_ want to be doing any types of spatial operations on `GeoDataFrames` that aren't in the same spatial reference, but it also seems we can get just this check without the heavy runtime hit of `try` / `except`. I didn't realize they were such a perf hit 😱

Comment (SybelBlue) [reply src/synthesize.py]: ~I've been reviewing those lines and I realized another reason why I switched - it often reports that nothing is equal. I've been reading it and messing with it and I haven't isolated why, just that it performs and behaves completely different from `GeoDataFrame.equals`. In fact, reverting the code to the old equality checker makes the synthesizer instantly fail the univariate synthesis task...~

Silly mistake on my part, sorry. Let me get back to you.

File: src/example.py (hunk 3 above)
Comment (SybelBlue) [reply src/example.py:45]: Got it!

File: src/synthesize.py (hunk 4 above)
Comment (SybelBlue) [reply src/synthesize.py]: I tried to adapt that method by delaying the use of try/except as much as possible. It worked, but it's ugly, and the performance gains are minor, less than 10%. Let me know if you want me to go back and just use one try/except around `assert_geoframe_equals`. Otherwise we should be good to go!

Comment (parkerziegler) [reply src/synthesize.py]: I don't think it's worth the perf optimization to bring in all of this code at this stage, so let's stick with `assert_geodataframe_equal` for now, and in the long run of this project I'll know to focus on this area for optimization.
